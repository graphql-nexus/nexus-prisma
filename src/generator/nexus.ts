import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import * as fs from 'fs-jetpack'
import * as OS from 'os'
import { GeneratedModule } from './types'

/** Transform the given DMMF into TS source code. */
export function generateRuntimeNexusSource(dmmf: DMMF.Document): GeneratedModule {
  if (process.env.NP_DEBUG) {
    fs.write('dmmf.json', dmmf)
  }

  const content = endent`
    import * as Nexus from 'nexus'

    ${dmmf.datamodel.models.map(modelToTypeScriptSource).join(OS.EOL + OS.EOL)}
  `

  return {
    fileName: 'nexus.ts',
    content,
  }
}

/** Transform a DMMF model into TS source code. */
function modelToTypeScriptSource(model: DMMF.Model): string {
  return endent`

    export const ${model.name} = {
      ${model.fields.filter(isSupportedType).map(fieldToTypeScriptSource).join(`,${OS.EOL}`)},
    } as const
  `
}

/**
 * Transform a DMMF model _field_ into TS source code.
 *
 * @example
 *   // Example of field data given:
 *   const field = {
 *     name: 'f16',
 *     kind: 'scalar',
 *     isList: false,
 *     isRequired: true,
 *     isUnique: false,
 *     isId: false,
 *     isReadOnly: false,
 *     type: 'DateTime',
 *     hasDefaultValue: false,
 *     isGenerated: false,
 *     isUpdatedAt: false,
 *   }
 */
function fieldToTypeScriptSource(field: DMMF.Field) {
  return endent`
    ${field.name}: {
      name: '${field.name}',
      config: {
        type: ${renderType(field)},
      }
    }
  `
}

function renderType(field: DMMF.Field): string {
  const graphqlType = getGraphQLType(field)

  if (field.isList) {
    return endent`
      Nexus.nonNull(Nexus.list(Nexus.nonNull(${quote(graphqlType)})))
    `
  } else if (field.isRequired) {
    return endent`
      Nexus.nonNull(${quote(graphqlType)})
    `
  } else {
    return endent`
      Nexus.nullable(${quote(graphqlType)})
    `
  }
}

type PrismaScalarType =
  | 'String'
  | 'Boolean'
  | 'Int'
  | 'BigInt'
  | 'Float'
  | 'Decimal'
  | 'DateTime'
  | 'Json'
  | 'Bytes'

function quote(x: string): string {
  return `'${x}'`
}

type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

function isSupportedType(field: DMMF.Field): boolean {
  return field.kind === 'scalar'
}

function getGraphQLType(field: DMMF.Field): GraphQLScalarType {
  if (field.kind !== 'scalar') {
    throw new Error(`type mapping only supported for scalar-type fields`)
  }

  const typeName = field.type as PrismaScalarType

  switch (typeName) {
    case 'String': {
      if (field.isId) {
        return GraphQLScalarTypes.ID
      }
      return GraphQLScalarTypes.String
    }
    default: {
      return GraphQLScalarTypes.String
    }
  }
}

const GraphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
} as const
