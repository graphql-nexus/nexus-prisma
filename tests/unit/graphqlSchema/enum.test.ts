import dedent from 'dindist'
import { enumType } from 'nexus'
import { testGraphqlSchema } from '../../__helpers__/testers'

testGraphqlSchema({
  description: 'When an enum is defined in the Prisma schema it can be projected into the GraphQL API',
  datasourceSchema: dedent`
    enum Foo {
      a
    }
  `,
  apiSchema({ Foo }) {
    return [enumType(Foo)]
  },
})

testGraphqlSchema({
  description:
    'When an enum with multiple members is defined in the Prisma schema it and all its members can be projected into the GraphQL API',
  datasourceSchema: dedent`
    enum Foo {
      a
      b
      c
    }
  `,
  apiSchema({ Foo }) {
    return [enumType(Foo)]
  },
})

testGraphqlSchema({
  description:
    'When an enum is defined with documentation in the Prisma schema it can be projected into the GraphQL API with that documentation',
  datasourceSchema: dedent`
    /// Some documentation
    enum Foo {
      a
    }
  `,
  apiSchema({ Foo }) {
    return [enumType(Foo)]
  },
})
