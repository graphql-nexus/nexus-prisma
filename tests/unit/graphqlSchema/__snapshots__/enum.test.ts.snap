// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When an enum is defined in the Prisma schema it can be projected into the GraphQL API: graphqlSchema 1`] = `
"
enum Foo {
  a
}
"
`;

exports[`When an enum is defined with documentation in the Prisma schema it can be projected into the GraphQL API with that documentation: graphqlSchema 1`] = `
"
\\"\\"\\"Some documentation\\"\\"\\"
enum Foo {
  a
}
"
`;

exports[`When an enum with multiple members is defined in the Prisma schema it and all its members can be projected into the GraphQL API: graphqlSchema 1`] = `
"
enum Foo {
  a
  b
  c
}
"
`;
